lib/rubocop/cop/lint/non_deterministic_require_order.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Dir[...] and Dir.glob(...) do not make any guarantees about
+      # the order in which files are returned. The final order is
+      # determined by the operating system and file system.
+      # This means that using them in cases where the order matters,
+      # such as requiring files, can lead to intermittent failures
+      # that are hard to debug. To ensure this doesn't happen,
+      # always sort the list.
+      #
+      # @example
+      #
+      #   # bad
+      #   Dir["./lib/**/*.rb"].each do |file|
+      #     require file
+      #   end
+      #
+      #   # good
+      #   Dir["./lib/**/*.rb"].sort.each do |file|
+      #     require file
+      #   end
+      #
+      # @example
+      #
+      #   # bad
+      #   Dir.glob(Rails.root.join(__dir__, 'test', '*.rb')).each do |file|
+      #     require file
+      #   end
+      #
+      #   # good
+      #   Dir.glob(Rails.root.join(__dir__, 'test', '*.rb')).sort.each do |file|
+      #     require file
+      #   end
+      #
+      class NonDeterministicRequireOrder < Cop
+        MSG = 'Sort files before requiring them.'
+
+        def_node_matcher :unsorted_dir_loop?, <<~PATTERN
+          (send (send (const nil? :Dir) {:[] :glob} ...) :each)
+        PATTERN
+
+        def_node_matcher :loop_variable, <<~PATTERN
+          (args (arg $_))
+        PATTERN
+
+        def_node_search :var_is_required?, <<~PATTERN
+          (send nil? :require (lvar %1))
+        PATTERN
+
+        def on_block(node)
+          return unless node.children && node.children.length > 2
+
+          block_start = node.children[0]
+
+          return unless unsorted_dir_loop?(block_start)
+
+          var_name = loop_variable(node.children[1])
```

Comment:
Block nodes have some methods to access their parts that could make this more readable:
```ruby
return unless node.body

block_start = node.send_node
...
var_name = loop_variable(node.arguments)
```
See https://github.com/rubocop-hq/rubocop/blob/bcd21f9fe309310423cfb9cee39c979bab56768c/lib/rubocop/ast/node/block_node.rb#L16-L35

Reply:
Much nicer! It also makes me more confident that the array will look like what we expect here.